src/esp/assets/ResourceManager.cpp

```diff
@@ -2121,8 +2105,339 @@ bool compareShaderTypeToMnMatType(const ObjectInstanceShaderType typeToCheck,
+    // original material.
+    appendIfNotPresent(origMaterialData, newAttributes,
+                       MaterialAttribute::BaseColor, baseColor);
+    appendIfNotPresent(origMaterialData, newAttributes,
+                       MaterialAttribute::Metalness, metalness);
+
+    // if diffuse texture is present, use as base color texture in pbr.
+    if (phongMaterial.hasAttribute(MaterialAttribute::DiffuseTexture)) {
+      uint32_t bcTextureVal = phongMaterial.diffuseTexture();
+      appendIfNotPresent(origMaterialData, newAttributes,
+                         MaterialAttribute::BaseColorTexture, bcTextureVal);
+    }
+    // texture transforms, if there's none the returned matrix is an
+    // identity Only copy if we don't already have TextureMatrix attribute
+    // (if original phongMaterial does not have that specific attribute)
+    if (!setTexMatrix &&
+        (!phongMaterial.hasAttribute(MaterialAttribute::TextureMatrix))) {
+      arrayAppend(newAttributes, {MaterialAttribute::TextureMatrix,
+                                  phongMaterial.commonTextureMatrix()});
+      // setTexMatrix = true;
+    }
+
+    // base texture
+
+  }  // if no PbrMetallicRoughness material support exists in material
+
+  // build flags to support all materials
+  constexpr auto flags = Mn::Trade::MaterialType::Flat |
+                         Mn::Trade::MaterialType::Phong |
+                         Mn::Trade::MaterialType::PbrMetallicRoughness;
+
+  // create new material from attributes array
+  Mn::Trade::MaterialData newMaterialData{flags, std::move(newAttributes)};
+
+  return newMaterialData;
+}  // ResourceManager::createUniversalMaterial
+
 }  // namespace
 
+Mn::Trade::MaterialData ResourceManager::buildDefaultPhongMaterial(
+    Mn::Color4 amb,   // = Mn::Color4{0.1},
+    Mn::Color4 diff,  // = Mn::Color4{0.7 * 0.175},
+    Mn::Color4 spec,  // = Mn::Color4{0.2 * 0.175},
+    float shininess /*= 80.0f*/) {
+  Mn::Trade::MaterialData materialData{
+      Mn::Trade::MaterialType::Phong,
+      {{Mn::Trade::MaterialAttribute::AmbientColor, amb},
+       {Mn::Trade::MaterialAttribute::DiffuseColor, diff},
+       {Mn::Trade::MaterialAttribute::SpecularColor, spec},
+       {Mn::Trade::MaterialAttribute::Shininess, shininess}}};
+  return materialData;
+}
+
+Mn::Trade::MaterialData ResourceManager::setDefaultMaterialUserAttributes(
+    const Mn::Trade::MaterialData& material,
+    ObjectInstanceShaderType shaderTypeToUse,
+    bool hasVertObjID,
+    bool hasTxtrObjID,
+    int txtrIdx) const {
+  // New material's attributes
+  Cr::Containers::Array<Mn::Trade::MaterialAttributeData> newAttributes;
+  arrayAppend(newAttributes, Cr::InPlaceInit, "hasPerVertexObjectId",
+              hasVertObjID);
+  arrayAppend(newAttributes, Cr::InPlaceInit, "hasTextureObjectId",
+              hasTxtrObjID);
+
+  if (hasTxtrObjID) {
+    arrayAppend(newAttributes, Cr::InPlaceInit, "objectIdTexturePtr",
```

Comment:
I was actually using the Ptr label to differentiate it from the usual texture index, which otherwise is the same attribute name as this except for the first character.

Reply:
I don't see that as a practical problem, but I admit it's a matter of preference.